oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractWebClientReactiveOAuth2AccessTokenResponseClient.java

```diff
@@ -0,0 +1,229 @@
+	 */
+	protected abstract Set<String> scopesToRequest(T grantRequest);
+
+	/**
+	 * Returns the scopes to include in the response if the authorization
+	 * server returned no scopes in the response.
+	 *
+	 * <p>As per <a href="https://tools.ietf.org/html/rfc6749#section-5.1">RFC-6749 Section 5.1 Successful Access Token Response</a>,
+	 * if AccessTokenResponse.scope is empty, then default to the scope
+	 * originally requested by the client in the Token Request.</p>
+	 *
+	 * @param grantRequest the grant request
+	 * @return the scopes to include in the response if the authorization
+	 *         server returned no scopes.
+	 */
+	protected Set<String> defaultScopes(T grantRequest) {
+		return scopesToRequest(grantRequest);
+	}
+
+	/**
+	 * Reads the token response from the response body.
+	 *
+	 * @param grantRequest the request for which the response was received.
+	 * @param response the client response from which to read
+	 * @return the token response from the response body.
+	 */
+	private Mono<OAuth2AccessTokenResponse> readTokenResponse(T grantRequest, ClientResponse response) {
+		return response.body(oauth2AccessTokenResponse())
+				.onErrorMap(OAuth2AuthorizationException.class, e -> createClientAuthorizationException(
+						response,
+						clientRegistration(grantRequest).getRegistrationId(),
+						e))
+				.map(tokenResponse -> populateTokenResponse(grantRequest, tokenResponse));
+	}
+
+	/**
+	 * Wraps the given {@link OAuth2AuthorizationException} in a {@link ClientAuthorizationException}
+	 * that provides response details, and a more descriptive exception message.
+	 *
+	 * @param response the token response
+	 * @param clientRegistrationId the id of the {@link ClientRegistration} for which a token is being requested
+	 * @param authorizationException the {@link OAuth2AuthorizationException} to wrap
+	 * @return the {@link ClientAuthorizationException} that wraps the given {@link OAuth2AuthorizationException}
+	 */
+	private OAuth2AuthorizationException createClientAuthorizationException(
+			ClientResponse response,
+			String clientRegistrationId,
+			OAuth2AuthorizationException authorizationException) {
+
+		String message = String.format("Error retrieving OAuth 2.0 Access Token (HTTP Status Code: %s) %s",
+				response.rawStatusCode(),
+				authorizationException.getError());
+
+		return new ClientAuthorizationException(
+				authorizationException.getError(),
+				clientRegistrationId,
+				message,
+				authorizationException);
+	}
+
+	/**
+	 * Populates the given {@link OAuth2AccessTokenResponse} with additional details
+	 * from the grant request.
+	 *
+	 * @param grantRequest the request for which the response was received.
+	 * @param tokenResponse the original token response
+	 * @return a token response optionally populated with additional details from the request.
+	 */
+	protected OAuth2AccessTokenResponse populateTokenResponse(T grantRequest, OAuth2AccessTokenResponse tokenResponse) {
```

Comment:
It's overridden by WebClientReactiveRefreshTokenTokenResponseClient, which has slightly different semantics.

As an alternative, I'm going to make this method, and the other protected methods use default visibility instead of protected, so they can't be utilized outside of the package